### PR TITLE
fix: token refresh locking logic

### PR DESCRIPTION
**Pull request by TBonnin**
One customer is experiencing connection refresh failures with the following error:
Acquiring lock for key: lock:refresh:XXX timed out after 12000ms They are making concurrent calls to `/proxy` while the token is expired. Each request trying to acquire the lock is waiting and then attempts again to refresh the token. When a lot of concurrent requests, waiting for all the other requests in the queue to refetch the connection token can exceed the lock acquisition timeout.
This commit is introducing 2 fixes:
- after waiting for acquiring the lock, first check if the connection has been refreshed, if yes returns the credentials directly.
- if lock acquisition times out, also check if the connection has been refreshed and only fails if it is not the case.

## Issue ticket number and link
https://linear.app/nango/issue/NAN-1164/salesforcebox-connections-failing-to-refresh-intermittently

## Checklist before requesting a review (skip if just adding/editing APIs & templates)
- [ ] I added tests, otherwise the reason is: 
- [ ] I added observability, otherwise the reason is:
- [ ] I added analytics, otherwise the reason is:

### CODE REVIEW

File: packages/shared/lib/services/connection.service.ts [hunk 1]
```diff
@@ -881,18 +882,70 @@ class ConnectionService {
...
+                                success: true,
+                                error: null,
+                                response: {
+                                    refreshed: true,
+                                    credentials: connection.credentials as OAuth2Credentials | AppCredentials | AppStoreCredentials | OAuth2ClientCredentials
+                                }
+                            };
+                        }
+                    }
+                    throw err;
```
Comment (TBonnin) [inline packages/shared/lib/services/connection.service.ts]: there is some code duplication but it is not trivial to consolidate the logci into one function. Let me know if you think I should spend more time trying to be more DRY

Comment (khaliqgant) [reply packages/shared/lib/services/connection.service.ts]: I find this logic very confusing and hard to follow. This method has always been hard to follow IMO but if possible it would be great to make it more readable / easier to follow. Definitely push back if you think it would be too much of a time sink though

**Review by khaliqgant — APPROVED** [review]
Fix makes sense, but the logic is hard to follow. See my feedback inline

File: packages/shared/lib/services/connection.service.ts (hunk 1 above)
Comment (TBonnin) [reply packages/shared/lib/services/connection.service.ts]: The locking logic is indeed complicated. My quick attempt at trying to abstract the complexity was not very successful though, I was basically just moving the complexity around. I am happy to pair if you have some idea about how to improve the locking logic

Comment (TBonnin) [reply packages/shared/lib/services/connection.service.ts]: Made an attempt at removing the duplication [7ad0229](https://github.com/NangoHQ/nango/pull/2494/commits/7ad022989373db94678aa34db4de5b8b616c8a8e)
The concurrency/locking logic is still the same though

File: packages/shared/lib/services/connection.service.ts [hunk 2]
```diff
@@ -846,150 +847,151 @@ class ConnectionService {
...
+        if (!shouldRefresh) {
+            return {
+                success: true,
+                error: null,
+                response: {
+                    refreshed: false,
+                    credentials: connection.credentials as OAuth2Credentials | AppCredentials | AppStoreCredentials | OAuth2ClientCredentials
+                }
+            };
+        }
```
Comment (bodinsamuel) [inline packages/shared/lib/services/connection.service.ts]: this early call/return is incorrect, if a refresh is in progress the credentials can be invalid just after the return

**Review by bodinsamuel — COMMENTED** [review]
Agreed with Khaliq it's hard to read, maybe there is an alternative with a recursive call (not a good idea) or a while loop.

Also I think my comment about the early return could explain some bugs, we should always wait for lock and re-fetch the connection  .

```ts
# pseudo code
const lockKey = `lock:refresh:${environment_id}:${providerConfigKey}:${connectionId}`;
const ttlInMs = 10000;
const acquisitionTimeoutMs = ttlInMs * 1.2; 
let maxTries = 0;
try {
    do {
        try {
            await this.locking.tryAcquire(lockKey, ttlInMs, acquisitionTimeoutMs);
        } catch (err) {
            report(err);
            maxTries += 1;
            continue;
        }

        const { connection, shouldRefresh } = await getConnection();
        if (!shouldRefresh) {
            return { success: true, error: null, response: { refreshed: false, credentials: connection.credentials as any } };
        }

        const newCreds = await this.getNewCredentials(connection, providerConfig, template);
        return { success: true, error: null, response: { refreshed: true, credentials: newCreds } };
    } while (maxTries <= 10);
} finally {
    await this.locking.release(lockKey);
}
```

File: packages/shared/lib/services/connection.service.ts (hunk 2 above)
Comment (TBonnin) [reply packages/shared/lib/services/connection.service.ts]: good point. that's another problem that I haven't address, even though in practice I am not sure it can happen.

**Comment by TBonnin** [comment]
> Agreed with Khaliq it's hard to read, maybe there is an alternative with a recursive call (not a good idea) or a while loop.
> 
> Also I think my comment about the early return could explain some bugs, we should always wait for lock and re-fetch the connection .

I agree the early return might explain some issues. I have modified my PR to always acquire the lock and re-fetch the connection. 
I haven't added the loop because I feel it would add to much latency to retry several times. I also kept the attempt at checking the freshness of the credentials when the lock times out. It is still duplication but only twice :D [fce7cbe](https://github.com/NangoHQ/nango/pull/2494/commits/fce7cbe9cfe993cc9fc3f899f429da1b59e6067c)
Let me know what you think @bodinsamuel

**Comment by bodinsamuel** [comment]
Yes agreed for now, we know it can be improved like this but it has some drawbacks 👌🏻

**Review by bodinsamuel — APPROVED** [review]
